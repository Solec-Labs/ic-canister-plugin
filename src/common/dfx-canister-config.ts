import path from 'path';
import type { DfxNetwork } from './dfx-network';
import { loadJsonFile } from './load-json-file';

export type CanisterName = string;
export type CanisterId = string;

export type CanisterNetworkConfig = Record<DfxNetwork, CanisterId>;
export type CanisterConfig = Record<CanisterName, CanisterNetworkConfig>;

export function getCanisterConfig(network: DfxNetwork): CanisterConfig {
  const canisterConfigPath =
    network === 'ic'
      ? path.resolve('canister_ids.json')
      : path.resolve('.dfx', network, 'canister_ids.json');

  return loadJsonFile(canisterConfigPath);
}
