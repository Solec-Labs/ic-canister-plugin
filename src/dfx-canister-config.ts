import path from 'path';
import fs from 'fs';
import { IcWebpackPluginError } from './ic-webpack-plugin-error';
import type { NetworkName } from './dfx-network';

export type CanisterName = string;
export type CanisterId = string;

export type DfxCanisterNetworkConfig = Record<NetworkName, CanisterId>;
export type DfxCanisterConfig = Record<CanisterName, DfxCanisterNetworkConfig>;

export function getCanisterConfig(network: NetworkName): DfxCanisterConfig {
  const canisterConfigPath =
    network === 'ic'
      ? path.resolve('canister_ids.json')
      : path.resolve('.dfx', network, 'canister_ids.json');

  let canisterConfigBuffer: Buffer | undefined;
  try {
    canisterConfigBuffer = fs.readFileSync(canisterConfigPath);
  } catch (error) {
    throw new IcWebpackPluginError(
      `Error loading ${canisterConfigPath}`,
      error,
    );
  }

  const canisterConfigString = canisterConfigBuffer.toString();

  let canisterConfig: DfxCanisterConfig | undefined;
  try {
    canisterConfig = JSON.parse(canisterConfigString) as DfxCanisterConfig;
  } catch (error) {
    throw new IcWebpackPluginError(
      `Error parsing the content of ${canisterConfigPath} into JSON`,
      error,
    );
  }

  return canisterConfig;
}
