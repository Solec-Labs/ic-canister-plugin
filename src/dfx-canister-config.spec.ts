import * as fs from 'fs';
import * as path from 'path';
import { DfxCanisterConfig, getCanisterConfig } from './dfx-canister-config';

jest.mock('fs');
type ReadFileSync = typeof fs['readFileSync'];

describe('getCanisterConfig', () => {
  let readFileSyncMock: jest.SpyInstance<
    ReturnType<ReadFileSync>,
    Parameters<ReadFileSync>
  >;

  const canisterConfig: DfxCanisterConfig = {
    canister_frontend: {
      local: 'rrkah-fqaaa-aaaaa-aaaaq-cai',
      testnet: 'ryjl3-tyaaa-aaaaa-aaaba-cai',
      ic: 'r7inp-6aaaa-aaaaa-aaabq-cai',
    },
    canister_backend: {
      local: 'rkp4c-7iaaa-aaaaa-aaaca-cai',
      testnet: 'renrk-eyaaa-aaaaa-aaada-cai',
      ic: 'rno2w-sqaaa-aaaaa-aaacq-cai',
    },
  };
  const canisterConfigString = JSON.stringify(canisterConfig);
  const canisterConfigBuffer = Buffer.from(canisterConfigString);

  beforeEach(() => {
    readFileSyncMock = jest.spyOn(fs, 'readFileSync');
  });

  it('should get the ic network from the root folder', () => {
    readFileSyncMock.mockReturnValue(canisterConfigBuffer);

    const result = getCanisterConfig('ic');

    expect(readFileSyncMock).toHaveBeenCalledWith(
      path.resolve('canister_ids.json'),
    );
    expect(result).toEqual(canisterConfig);
  });

  it.each(['local', 'testnet'])(
    'should get the %s network config from the .dfx folder',
    network => {
      readFileSyncMock.mockReturnValue(canisterConfigBuffer);

      const result = getCanisterConfig(network);

      expect(readFileSyncMock).toHaveBeenCalledWith(
        path.resolve('.dfx', network, 'canister_ids.json'),
      );
      expect(result).toEqual(canisterConfig);
    },
  );

  it('should throw if it cannot read the canister path', () => {
    readFileSyncMock.mockImplementation(() => {
      throw new Error('Path does not exist');
    });

    expect(() => getCanisterConfig('ic')).toThrowErrorMatchingSnapshot();
  });

  it('should throw if the file buffer cannot be decoded to JSON', () => {
    readFileSyncMock.mockReturnValue(Buffer.from(''));

    expect(() => getCanisterConfig('ic')).toThrowErrorMatchingSnapshot();
  });
});
