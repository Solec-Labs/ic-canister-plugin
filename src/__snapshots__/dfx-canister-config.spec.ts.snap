// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCanisterConfig should throw if it cannot read the canister path 1`] = `"Error loading /home/nathan/programming/crypto/ic-webpack-plugin/canister_ids.json"`;

exports[`getCanisterConfig should throw if the file buffer cannot be decoded to JSON 1`] = `"Error parsing the content of /home/nathan/programming/crypto/ic-webpack-plugin/canister_ids.json into JSON"`;
